mat64/hogsvd.go

```diff
@@ -0,0 +1,180 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+// Based on the SingularValueDecomposition class from Jama 1.0.3.
+
+package mat64
+
+import (
+	"errors"
+
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/matrix"
+)
+
+// HOGSVD is a type for creating and using the Higher Order Generalized Singular Value
+// Decomposition (HOGSVD) of a set of matrices.
+type HOGSVD struct {
+	n int
+	v *Dense
+	b []Dense
+
+	err error
+}
+
+// Factorize computes the higher order generalized singular value decomposition (GSVD)
+// of the n input r_i×c column matrices in m.
+//
+//  M_0 = U_0 * Σ_0 * V^T
+//  M_1 = U_1 * Σ_1 * V^T
+//  .
+//  .
+//  .
+//  M_{n-1} = U_{n-1} * Σ_{n-1} * V^T
+//
+// where U_i are r_i×c matrices of singular vectors, Σ are c×c matrices singular values, and V
+// is a c×c matrix of singular vectors.
+//
+// Factorize returns whether the decomposition succeeded. If the decomposition
+// failed, routines that require a successful factorization will panic.
+func (gsvd *HOGSVD) Factorize(m ...Matrix) (ok bool) {
+	if len(m) < 2 {
+		panic("hogsvd: too few matrices")
+	}
+	gsvd.n = 0
+
+	r, c := m[0].Dims()
+	a := make([]Dense, len(m))
+	for i, d := range m {
+		rd, cd := d.Dims()
+		if rd > r {
+			r = rd
+		}
+		if cd != c {
+			panic(matrix.ErrShape)
+		}
+		a[i].Mul(d.T(), d)
+	}
+
+	s := getWorkspace(c, c, true)
+	defer putWorkspace(s)
+	sij := getWorkspace(c, c, false)
+	defer putWorkspace(sij)
+	for i, ai := range a {
+		for _, aj := range a[i+1:] {
+			gsvd.err = sij.Solve(ai.T(), aj.T())
```

Comment:
`a[i] = d^T * d`, so a is a positive definite matrix. It seems like this would be better all around if this used the Cholesky decomposition for the solve. For instance, compute all the choleskies before the a_i range, and so the "inverses" are cached for both Solves. This alone should cut out a bunch of N^3 steps. There may also be something tricky about linear solves with two positive definite matrices.